contentgrid-spring-boot-autoconfigure/src/main/java/com/contentgrid/spring/boot/autoconfigure/messaging/ContentGridPropertiesMessageConverter.java

```diff
@@ -0,0 +1,45 @@
+package com.contentgrid.spring.boot.autoconfigure.messaging;
+
+import lombok.RequiredArgsConstructor;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.converter.SmartMessageConverter;
+import org.springframework.messaging.support.MessageBuilder;
+
+/**
+ * Delegates message conversion to a composite MessageConverter, but adds applicationId and deploymentId to the headers
+ */
+@RequiredArgsConstructor
+public class ContentGridPropertiesMessageConverter implements SmartMessageConverter {
+
+    private final SmartMessageConverter delegate;
+    private final ContentGridMessagingSystemProperties systemProperties;
+
+    @Override
+    public Object fromMessage(Message<?> message, Class<?> targetClass, Object conversionHint) {
+        return delegate.fromMessage(message, targetClass, conversionHint);
+    }
+
+    @Override
+    public Message<?> toMessage(Object payload, MessageHeaders headers, Object conversionHint) {
+        return delegate.toMessage(payload, extendHeaders(headers), conversionHint);
+    }
+
+    @Override
+    public Object fromMessage(Message<?> message, Class<?> targetClass) {
+        return delegate.fromMessage(message, targetClass);
+    }
+
+    @Override
+    public Message<?> toMessage(Object payload, MessageHeaders headers) {
+        return delegate.toMessage(payload, extendHeaders(headers));
+    }
+
+    private MessageHeaders extendHeaders(MessageHeaders headers) {
+        return MessageBuilder.withPayload("foo").copyHeaders(headers)
+                .setHeaderIfAbsent("applicationId", systemProperties.getApplicationId())
+                .setHeaderIfAbsent("deploymentId", systemProperties.getDeploymentId())
+                .build()
+                .getHeaders();
```

Comment:
`MessageHeaders` is mutated everywhere; so why the effort to make them immutable?

Reply:
From the javadoc of MessageHeaders:
> **IMPORTANT**: This class is immutable. Any mutating operation such as `put(..)`, `putAll(..)` and others will throw `UnsupportedOperationException`.
> Subclasses do have access to the raw headers, however, via `getRawHeaders()`.
> One way to create message headers is to use the `MessageBuilder`:
> `MessageBuilder.withPayload("foo").setHeader("key1", "value1").setHeader("key2", "value2");`
> ...

What way to mutate it would you like me to use?

Reply:
There are 2 options

- `MessageHeaders` implements `Map`, so you could copy that to a mutable `Map`, insert the additional headers and then create a new `MessageHeaders` from that map.
- Use the `MessageHeaderAccessor` to mutate the headers and return a new `MessageHeaders`.